diff --git a/options/base_options.py b/options/base_options.py
index eba63f6..51b03b3 100644
--- a/options/base_options.py
+++ b/options/base_options.py
@@ -20,7 +20,7 @@ class BaseOptions:
     def initialize(self, parser):
         """Define the common options that are used in both training and test."""
         # basic parameters
-        parser.add_argument("--dataroot", required=True, help="path to images (should have subfolders trainA, trainB, valA, valB, etc)")
+        parser.add_argument("--dataroot", help="path to images (should have subfolders trainA, trainB, valA, valB, etc)")
         parser.add_argument("--name", type=str, default="experiment_name", help="name of the experiment. It decides where to store samples and models")
         parser.add_argument("--checkpoints_dir", type=str, default="./checkpoints", help="models are saved here")
         # model parameters
@@ -37,7 +37,7 @@ class BaseOptions:
         parser.add_argument("--init_gain", type=float, default=0.02, help="scaling factor for normal, xavier and orthogonal.")
         parser.add_argument("--no_dropout", action="store_true", help="no dropout for the generator")
         # dataset parameters
-        parser.add_argument("--dataset_mode", type=str, default="unaligned", help="chooses how datasets are loaded. [unaligned | aligned | single | colorization]")
+        parser.add_argument("--dataset_mode", type=str, default="unaligned", help="chooses how datasets are loaded. [unaligned | aligned | single | colorization | pickle]")
         parser.add_argument("--direction", type=str, default="AtoB", help="AtoB or BtoA")
         parser.add_argument("--serial_batches", action="store_true", help="if true, takes images in order to make batches, otherwise takes them randomly")
         parser.add_argument("--num_threads", default=4, type=int, help="# threads for loading data")
diff --git a/test.py b/test.py
index 22fb6a4..662ebfd 100644
--- a/test.py
+++ b/test.py
@@ -32,7 +32,7 @@ from pathlib import Path
 from options.test_options import TestOptions
 from data import create_dataset
 from models import create_model
-from util.visualizer import save_images
+from util.visualizer import save_images, save_images_custom
 from util import html
 import torch
 
@@ -76,4 +76,5 @@ if __name__ == "__main__":
         if i % 5 == 0:  # save images to an HTML file
             print(f"processing ({i:04d})-th image... {img_path}")
         save_images(webpage, visuals, img_path, aspect_ratio=opt.aspect_ratio, width=opt.display_winsize)
-    webpage.save()  # save the HTML
+        # save_images_custom(visuals)
+    # webpage.save()  # save the HTML
diff --git a/util/util.py b/util/util.py
index f08c67b..5f8ddf9 100644
--- a/util/util.py
+++ b/util/util.py
@@ -6,6 +6,7 @@ import numpy as np
 from PIL import Image
 from pathlib import Path
 import torch.distributed as dist
+import pickle
 import os
 
 
@@ -24,7 +25,34 @@ def tensor2im(input_image, imtype=np.uint8):
         image_numpy = image_tensor[0].cpu().float().numpy()  # convert it into a numpy array
         if image_numpy.shape[0] == 1:  # grayscale to RGB
             image_numpy = np.tile(image_numpy, (3, 1, 1))
+            # image_numpy = image_numpy[0]
         image_numpy = (np.transpose(image_numpy, (1, 2, 0)) + 1) / 2.0 * 255.0  # post-processing: tranpose and scaling
+        # Denormalize and scale
+        # image_numpy = (image_numpy+ 1) / 2.0 * 255.0
+    else:  # if it is a numpy array, do nothing
+        image_numpy = input_image
+    return image_numpy.astype(imtype)
+
+def tensor2im_custom(input_image, imtype):
+    """ "Converts a Tensor array into a numpy image array.
+
+    Parameters:
+        input_image (tensor) --  the input image tensor array
+        imtype (type)        --  the desired type of the converted numpy array
+    """
+    if not isinstance(input_image, np.ndarray):
+        if isinstance(input_image, torch.Tensor):  # get the data from a variable
+            image_tensor = input_image.data
+        else:
+            return input_image
+        # print(image_tensor[0][0].detach())
+        image_numpy = image_tensor[0].cpu().float().numpy()  # convert it into a numpy array
+        if image_numpy.shape[0] == 1:  # grayscale to RGB
+            # image_numpy = np.tile(image_numpy, (3, 1, 1))
+            image_numpy = image_numpy[0]
+        # image_numpy = (np.transpose(image_numpy, (1, 2, 0)) + 1) / 2.0 * 255.0  # post-processing: tranpose and scaling
+        # Denormalize and scale
+        image_numpy = (image_numpy + 1) / 2.0 * 255
     else:  # if it is a numpy array, do nothing
         image_numpy = input_image
     return image_numpy.astype(imtype)
@@ -85,6 +113,7 @@ def save_image(image_numpy, image_path, aspect_ratio=1.0):
 
     image_pil = Image.fromarray(image_numpy)
     h, w, _ = image_numpy.shape
+    # h, w = image_numpy.shape
 
     if aspect_ratio > 1.0:
         image_pil = image_pil.resize((h, int(w * aspect_ratio)), Image.BICUBIC)
@@ -92,6 +121,20 @@ def save_image(image_numpy, image_path, aspect_ratio=1.0):
         image_pil = image_pil.resize((int(h / aspect_ratio), w), Image.BICUBIC)
     image_pil.save(image_path)
 
+def save_pickle_images(images_numpy, dataset_name):
+    """
+    Save numpy array of images to disk using pickle
+    """
+    # print(images_numpy.shape)
+    with open(f"{dataset_name}.pkl", "wb") as f:
+        pickle.dump(images_numpy, f)
+
+def save_pickle_image(image_numpy, image_name):
+    """
+    Save an images to disk using pickle
+    """
+    with open(f"{image_name}.pkl", "wb") as f:
+        pickle.dump(image_numpy, f)
 
 def print_numpy(x, val=True, shp=False):
     """Print the mean, min, max, median, std, and size of a numpy array
@@ -127,4 +170,4 @@ def mkdir(path):
     Parameters:
         path (str) -- a single directory path
     """
-    Path(path).mkdir(parents=True, exist_ok=True)
+    Path(path).mkdir(parents=True, exist_ok=True)
\ No newline at end of file
diff --git a/util/visualizer.py b/util/visualizer.py
index a3573fa..67b67c3 100644
--- a/util/visualizer.py
+++ b/util/visualizer.py
@@ -7,6 +7,7 @@ from pathlib import Path
 import wandb
 import os
 import torch.distributed as dist
+import pickle
 
 
 def save_images(webpage, visuals, image_path, aspect_ratio=1.0, width=256):
@@ -36,7 +37,28 @@ def save_images(webpage, visuals, image_path, aspect_ratio=1.0, width=256):
         links.append(image_name)
     webpage.add_images(ims, txts, links, width=width)
 
+def save_images_custom(visuals):
+    """My custom loginc to Save all images to the disk in a single pickel file.
 
+    Parameters:
+        webpage (the HTML class) -- the HTML webpage class that stores these imaegs (see html.py for more details)
+        visuals (OrderedDict)    -- an ordered dictionary that stores (name, images (either tensor or numpy) ) pairs
+        image_path (str)         -- the string is used to create image paths
+        aspect_ratio (float)     -- the aspect ratio of saved images
+        width (int)              -- the images will be resized to width x width
+
+    This function will save images stored in 'visuals' to the HTML file specified by 'webpage'.
+    """
+    # name = Path(image_path[0]).stem
+    # webpage.add_header(name)
+    images = []
+    for _, im_data in visuals.items():
+        image = util.tensor2im_custom(im_data, imtype=np.float64)
+        images.append(image)
+    images = np.array(images)
+    print(images.shape)
+    util.save_pickle_images(images, "era5_downscaled")
+    
 class Visualizer:
     """This class includes several functions that can display/save images and print/save logging information.
 
@@ -165,4 +187,4 @@ class Visualizer:
         # Only save to log file on main process (rank 0)
         if local_rank == 0:
             with open(self.log_name, "a") as log_file:
-                log_file.write(f"{message}\n")  # save the message
+                log_file.write(f"{message}\n")  # save the message
\ No newline at end of file
